design/hostclaim-multitenancy.md

```diff
@@ -0,0 +1,497 @@
+  annotations in the reserved ``host.metal3.io`` domain.
+* When I destroy the host, the association is broken and another user can take
+  over the server.
+
+##### Server administrator point of view
+
+As the owner of the bare metal servers, I would like to easily control the users
+that can use my servers.
+The BareMetalHost resource is extended with a field ``hostClaimNamespaces`` with
+two optional subfields:
+
+* ``matches``: the content of this field is a regular expression
+  (in Golang re2 syntax). The name of the namespace where the HostClaim is located
+  must match this regular expression.
+* ``withLabel``: the content of the field is a list of strings representing
+  labels. The namespace of the HostClaim must have a label belonging to the
+  list.
+
+#### Multi-tenancy
+
+As an infrastructure administrator I would like to host several isolated
+clusters.
+
+All the servers in the data-center are registered as BareMetalHost in one or
+several namespaces under the control of the infrastructure manager. Namespaces
+are created for each tenant of the infrastructure. They create
+standard cluster definitions in those namespaces.
+
+When the cluster is started, a HostClaim is created for each Metal3Machine
+associated to the cluster. The ``hostSelector`` is inherited from the
+Metal3Machine. As in the original workflow, it is used to choose the BareMetalHost
+associated with the cluster, but he associated BareMetalHost is not in the same
+namespace as the HostClaim. The exact definition of the BareMetalHost remains
+hidden from the cluster user, only parts of its status and metadata are copied
+back to the HostClaim resource. With this information,
+the data template controller has enough details to compute the different
+secrets (userData, metaData and networkData) associated to the Metal3Machine.
+Those secrets are linked to the HostClaim and, ultimately, to the
+BareMetalHost.
+
+When the cluster is modified, new Machine and Metal3Machine resources replace
+the previous ones. The HostClaims follow the life-cycle of the Metal3Machines
+and are destroyed with them. The BareMetalHosts are recycled and are bound to
+new HostClaims, potentially belonging to other clusters.
+
+#### Compute Resources in Another Cluster
+
+As a user I would like to describe my cluster within a specific management
+cluster. However the resources I intend to use (such as BareMetalHosts or
+KubeVirt virtual machines) will be defined in a separate cluster.
+
+The multi-tenancy extension for BareMetalHost is extended to HostClaims.
+Metal3Machine field ``identityRef`` points to a
+secret containing a kubeconfig object. This kubeconfig will be utilized instead
+of the HostClaim controller service account to create and manage HostClaim
+resources
+
+```yaml
+  apiVersion: metal3.io/v1alpha1
+  kind: HostClaim
+  metadata:
+    name: host-claim-yyyy
+    namespace: user1-ns
+  spec:
+    kind: baremetalhost
+    credentials: bmh-cluster-credentials
+    userData:
```

Comment:
Let's group `userData`, `image` and co under something like `template` for the UX benefit (similarity to Deployments, etc)